pennylane_lightning_gpu/src/util/MPIManager.hpp

```diff
@@ -0,0 +1,612 @@
+#pragma once
+
+#include <algorithm>
+#include <bit>
+#include <complex>
+#include <cstring>
+#include <cuda.h>
+#include <cuda_runtime.h>
+#include <mpi.h>
+#include <stdexcept>
+#include <string>
+#include <typeindex>
+#include <typeinfo>
+#include <unordered_map>
+#include <vector>
+
+#include "Error.hpp"
+
+namespace Pennylane::MPI {
+inline void errhandler(int errcode, const char *str) {
+    char msg[MPI_MAX_ERROR_STRING];
+    int resultlen;
+    MPI_Error_string(errcode, msg, &resultlen);
+    fprintf(stderr, "%s: %s\n", str, msg);
+    MPI_Abort(MPI_COMM_WORLD, 1);
+}
+
+#define PL_MPI_IS_SUCCESS(fn)                                                  \
+    {                                                                          \
+        int errcode;                                                           \
+        errcode = (fn);                                                        \
+        if (errcode != MPI_SUCCESS)                                            \
+            errhandler(errcode, #fn);                                          \
+    }
+
+template <typename T> auto cppTypeToString() -> const std::string {
+    const std::string typestr = std::type_index(typeid(T)).name();
+    return typestr;
+}
+
+/**
+ * @brief MPI operation class. Maintains MPI related operations.
+ */
+class MPIManager {
+  private:
+    bool isExternalComm_;
+    size_t rank_;
+    size_t size_per_node_;
+    size_t size_;
+    MPI_Comm communicator_;
+
+    std::string vendor_;
+    size_t version_;
+    size_t subversion_;
+
+  public:
+    MPIManager() : communicator_(MPI_COMM_WORLD) {
+        isExternalComm_ = true;
+        int rank_int;
+        int size_int;
+        PL_MPI_IS_SUCCESS(MPI_Comm_rank(communicator_, &rank_int));
+        PL_MPI_IS_SUCCESS(MPI_Comm_size(communicator_, &size_int));
+
+        rank_ = static_cast<size_t>(rank_int);
+        size_ = static_cast<size_t>(size_int);
+
+        findVendor();
+        findVersion();
```

Comment:
I think it may be a better idea to separate out finding the vendor, from setting it. Maybe the method returns a value that we use to set `vendor_` instead?
The same with `findVersion()`, `getNumProcsPerNode()` and `check_mpi_config()`. Otherwise, if you wish to keep the functionality as is, make those methods private, as they shouldn't be accessible in the public API.